src/job_queue/queue.rs

```diff
@@ -307,30 +124,228 @@ impl<P: Ord + Send + Sync + 'static> JobQueue<P> {
+        let (next_job, num_pending) = {
+            // acquire mutex lock
+            let mut guard = shared_queue.lock().unwrap();
+
+            // pick the highest priority job
+            guard
+                .jobs
+                .make_contiguous()
+                .sort_by(|a, b| b.priority.cmp(&a.priority));
+            let job = guard.jobs.pop_front().unwrap();
+
+            // set highest priority job as the current job
+            guard.current_job = Some(CurrentJob {
+                job_num,
+                job_id: job.job_id,
+                cancel_tx: job.cancel_tx.clone(),
+            });
+
+            (job, guard.jobs.len())
+        }; // mutex lock is released when guard drops.
+
+        // log that we are starting a job
+        tracing::info!(
+            "  *** JobQueue: begin job #{} - {} - {} queued job(s) ***",
+            job_num,
+            next_job.job_id,
+            num_pending
+        );
+
+        // record time that job starts
+        let timer = tokio::time::Instant::now();
+
+        // spawn task that performs the job, either async or blocking.
+        let task_handle = if next_job.job.is_async() {
+            tokio::spawn(
+                async move { next_job.job.run_async_cancellable(next_job.cancel_rx).await },
+            )
+        } else {
+            tokio::task::spawn_blocking(move || next_job.job.run(next_job.cancel_rx))
+        };
+
+        let task_result = tokio::select! {
+            // handle msg over 'stop' channel which indicates we must exit the loop.
+            _ = rx_stop.changed() => {
+                tracing::debug!("task process_jobs received Stop message.");
+
+                // acquire mutex lock and obtain current_job info, if any.
+                let maybe_info = shared_queue.lock().unwrap().current_job.as_ref().map(|cj| (cj.job_id, cj.cancel_tx.clone()) );
+
+                // if there is a presently executing job we need to cancel it
+                // and wait for it to complete.
+                if let Some((job_id, cancel_tx)) = maybe_info {
+                    match cancel_tx.send(()) {
+                        Ok(()) => {
+                            // wait for channel to close, indicating job has cancelled (or otherwise completed)
+                            tracing::debug!("JobQueue: notified current job {} to cancel.  waiting...", job_id);
+                            cancel_tx.closed().await;
+                            tracing::debug!("JobQueue: current job {} has cancelled.", job_id);
+                        }
+                        Err(e) => {
+                            tracing::warn!("could not send cancellation msg to current job {}. {}", job_id, e)
+                        }
+                    }
+                }
+
+                // exit loop, processing ends.
+                break;
+            },
+
+            task_result = task_handle => task_result,
```

Comment:
I think the flow of reading is less interrupted if this is the first line in the `select!`.

Reply:
yes,  this was a last minute change.  I think I might move the stop handling into a separate function to keep the flow cleaner here.

Reply:
moved stop handling into separate fn and reversed select branch order in 7fc22f78b47a08ffe0899442cb35affdf2babb67